README.md
<!--
*** Thanks for checking out this README Template. If you have a suggestion that would
*** make this better please fork the repo and create a pull request or simple open
*** an issue with the tag "enhancement".
*** Thanks again! Now go create something AMAZING! :D
-->

<!-- PROJECT SHIELDS -->
<!--
*** I'm using markdown "reference style" links for readability.
*** Reference links are enclosed in brackets [ ] instead of parentheses ( ).
*** See the bottom of this document for the declaration of the reference variables
*** for build-url, contributors-url, etc. This is an optional, concise syntax you may use.
*** https://www.markdownguide.org/basic-syntax/#reference-style-links
-->
[![NPM](https://img.shields.io/npm/v/react-ion-slider.svg)](https://www.npmjs.com/package/react-ion-slider)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

<!-- PROJECT LOGO -->
<br />
<p align="center">
  <a href="https://github.com/gipyeong-lee/react-ion-slider"><h3 align="center">react-ion-slider</h3></a>
    <a href="http://ionden.com/a/plugins/ion.rangeSlider/api.html"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://gipyeong-lee.github.io/react-ion-slider/">View Demo</a>
    ·
    <a href="https://github.com/gipyeong-lee/react-ion-slider/issues">Report Bug</a>
    ·
  <a href="https://github.com/gipyeong-lee/react-ion-slider/issues">Request Feature</a>
 </p>

<!-- TABLE OF CONTENTS -->
## Table of Contents

- [Table of Contents](#table-of-contents)
- [About The Project](#about-the-project)
- [Getting Started](#getting-started)
  - [Prerequisites](#prerequisites)
  - [Installation](#installation)
- [Usage](#usage)
  - [1. Before using `react-ion-slider` please import module](#1-before-using-react-ion-slider-please-import-module)
  - [2. Implement below codes in the react render method](#2-implement-below-codes-in-the-react-render-method)
  - [3. Update options manually without componentDidMount](#3-update-options-manually-without-componentdidmount)
  - [1. Before using `react-ion-slider` please import jquery module in your HEAD](#1-before-using-react-ion-slider-please-import-jquery-module-in-your-head)
  - [2. Please load using dynamic import](#2-please-load-using-dynamic-import)
  - [Supported Props](#supported-props)
  - [Ref Functions](#ref-functions)
- [Roadmap](#roadmap)
- [Contributing](#contributing)
- [License](#license)
- [Contact](#contact)
- [Reference](#reference)

<!-- ABOUT THE PROJECT -->
## About The Project

Ion-rangeslider react module, this module mapping `ion-rangeslider` properties to props.

<!-- GETTING STARTED -->
## Getting Started

This is an example of how you may give instructions on setting up your project locally.
To get a local copy up and running follow these simple example steps.

### Prerequisites

This is an example of how to list things you need to use the software and how to install them.

- npm

```sh
npm install npm@latest -g
```

- react

```sh
npm install --save react react-dom
```

### Installation

1. Npm install

   ```sh
   npm install --save react-ion-slider
   ```

   ```
   yarn add react-ion-slider --save
   ```

<!-- USAGE EXAMPLES -->
## Usage

> Normal

### 1. Before using `react-ion-slider` please import module

```js
import IonRangeSlider from 'react-ion-slider'
```

### 2. Implement below codes in the react render method

```js
  <IonRangeSlider type={} min={} max={} from={} to={} step={} values={} keyboard={} />
```

### 3. Update options manually without componentDidMount

```js
<IonRangeSlider ref={r => this.ionSlider = r} />

this.ionSlider.update({skin: 'flat', min: 100, max: 500, from: 120, to: 240, type: 'double'})
```

> Next.js
### 1. Before using `react-ion-slider` please import jquery module in your HEAD

```js
<Head>
        <title>Create Next App</title>
        <link rel="icon" href="/favicon.ico" />
        <script src="https://code.jquery.com/jquery-3.5.1.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossOrigin="anonymous"></script>
</Head>
```

### 2. Please load using dynamic import

```js
const IonRangeSlider = dynamic(() => import('react-ion-slider'),{ssr: false})
```

### Supported Props

| Option                 | Defaults | Type     | Description                                                                                                                                                              |
|------------------------|----------|----------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Basic setup            |          |          |                                                                                                                                                                          |
| type                   | "single" | string   | Choose slider type, could be single - for one handle, or double for two handles                                                                                          |
| min                    | 10       | number   | Set slider minimum value                                                                                                                                                 |
| max                    | 100      | number   | Set slider maximum value                                                                                                                                                 |
| from                   | min      | number   | Set start position for left handle (or for single handle)                                                                                                                |
| to                     | max      | number   | Set start position for right handle                                                                                                                                      |
| Advanced setup         |          |          |                                                                                                                                                                          |
| step                   | 1        | number   | Set sliders step. Always > 0. Could be fractional.                                                                                                                       |
| values                 | []       | array    | Set up your own array of possible slider values. They could be numbers or strings. If the values array is set up, min, max and step param, are no longer can be changed. |
| keyboard               | TRUE     | boolean  | Activates keyboard controls. Move left: ←, ↓, A, S. Move right: →, ↑, W, D.                                                                                              |
| Grid setup             |          |          |                                                                                                                                                                          |
| grid                   | FALSE    | boolean  | Enables grid of values.                                                                                                                                                  |
| grid_margin            | TRUE     | boolean  | Set left and right grid borders.                                                                                                                                         |
| grid_num               | 4        | number   | Number of grid units.                                                                                                                                                    |
| grid_snap              | FALSE    | boolean  | Snap grid to sliders step (step param). If activated, grid_num will not be used.                                                                                         |
| Interval control       |          |          |                                                                                                                                                                          |
| drag_interval          | FALSE    | boolean  | Allow user to drag whole range. Only in double type                                                                                                                      |
| min_interval           | —        | number   | Set minimum diapason between sliders. Only in double type                                                                                                                |
| max_interval           | —        | number   | Set maximum diapason between sliders. Only in double type                                                                                                                |
| Handles control        |          |          |                                                                                                                                                                          |
| from_fixed             | FALSE    | boolean  | Fix position of left (or single) handle.                                                                                                                                 |
| from_min               | min      | number   | Set minimum limit for left handle.                                                                                                                                       |
| from_max               | max      | number   | Set the maximum limit for left handle                                                                                                                                    |
| from_shadow            | FALSE    | boolean  | Highlight the limits for left handle                                                                                                                                     |
| to_fixed               | FALSE    | boolean  | Fix position of right handle.                                                                                                                                            |
| to_min                 | min      | number   | Set the minimum limit for right handle                                                                                                                                   |
| to_max                 | max      | number   | Set the maximum limit for right handle                                                                                                                                   |
| to_shadow              | FALSE    | boolean  | Highlight the limits for right handle                                                                                                                                    |
| UI control             |          |          |                                                                                                                                                                          |
| skin                   | "flat"   | string   | Choose UI skin to use                                                                                                                                                    |
| hide_min_max           | FALSE    | boolean  | Hides min and max labels                                                                                                                                                 |
| hide_from_to           | FALSE    | boolean  | Hide from and to labels                                                                                                                                                  |
| force_edges            | FALSE    | boolean  | Slider will be always inside it's container.                                                                                                                             |
| extra_classes          | —        | string   | Traverse extra CSS-classes to slider container                                                                                                                           |
| block                  | FALSE    | boolean  | Locks slider and makes it inactive (visually). input is NOT disabled. Can still be send with forms.                                                                      |
| Prettify numbers       |          |          |                                                                                                                                                                          |
| prettify_enabled       | TRUE     | boolean  | Improve readability of long numbers. 10000000 → 10 000 000                                                                                                               |
| prettify_separator     | " "      | string   | Set up your own separator for long numbers. 10 000, 10.000, 10-000 etc.                                                                                                  |
| prettify               | null     | function | Set up your own prettify function. Can be anything. For example, you can set up unix time as slider values and than transform them to cool looking dates.                |
| —                      |          |          |                                                                                                                                                                          |
| prefix                 | —        | string   | Set prefix for values. Will be set up right before the number: $100                                                                                                      |
| postfix                | —        | string   | Set postfix for values. Will be set up right after the number: 100k                                                                                                      |
| max_postfix            | —        | string   | Special postfix, used only for maximum value. Will be showed after handle will reach maximum right position. For example 0 — 100+                                        |
| decorate_both          | TRUE     | boolean  | Used for "double" type and only if prefix or postfix was set up. Determine how to decorate close values. For example: $10k — $100k or $10 — 100k                         |
| values_separator       | " — "    | string   | Set your own separator for close values. Used for double type. Default: 10 — 100. Or you may set: 10 to 100, 10 + 100, 10 → 100 etc.                                     |
| Data control           |          |          |                                                                                                                                                                          |
| input_values_separator | " ; "    | string   | Separator for double values in input value property. Default FROM;TO. Only for double type                                                                               |
| disable                | FALSE    | boolean  | Locks slider and makes it inactive. inputis disabled too. Invisible to forms.                                                                                            |
| Callbacks              |          |          |                                                                                                                                                                          |
| scope                  | null     | object   | Scope for callbacks. Pass any object. Callback will be executed like this: onChange.call(scope);                                                                         |
| onStart                | null     | function | Callback. Is called on slider start.                                                                                                                                     |
| onChange               | null     | function | Callback. IS called on each values change.                                                                                                                               |
| onFinish               | null     | function | Callback. Is called than user releases handle.                                                                                                                           |
| onUpdate               | null     | function | Callback. Is called than slider is modified by external methods update or reset.                                                                                         |

### Ref Functions

| Option | Defaults | Type     | Description                     |
|--------|----------|----------|---------------------------------|
| update | null     | function | You can update options manually |

<!-- ROADMAP -->
## Roadmap

See the [open issues](https://github.com/gipyeong-lee/react-ion-slider/issues) for a list of proposed features (and known issues).

<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to be learn, inspire, and create. Any contributions you make are **greatly appreciated**.

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE` for more information.

<!-- CONTACT -->
## Contact

GiPyeong Lee - [@gipyeong](https://linkedin.com/in/gipyeong-lee-85734396) - gipyeong.lee@gmail.com

Project Link: [https://github.com/gipyeong-lee/react-ion-slider](https://github.com/gipyeong-lee/react-ion-slider)

## Reference

[Best Readme Template](https://github.com/othneildrew/Best-README-Template)


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [1.0.15](https://github.com/gipyeong-lee/react-ion-slider/compare/1.0.14...1.0.15)

- modified hooks [`6fae633`](https://github.com/gipyeong-lee/react-ion-slider/commit/6fae6336b9180a5bf1c656b66bb2ec98c93fcec0)

#### [1.0.14](https://github.com/gipyeong-lee/react-ion-slider/compare/1.0.13...1.0.14)

> 22 February 2021

- Release react-ion-slider 1.0.14 [`ef2647b`](https://github.com/gipyeong-lee/react-ion-slider/commit/ef2647b6a7f0b1ebd7c351657a6a7a2a67a0709c)
- implement after release sync master develop [`b5eb850`](https://github.com/gipyeong-lee/react-ion-slider/commit/b5eb8503a0c9bd824c63b32a71c73647f0b3b8d2)

#### [1.0.13](https://github.com/gipyeong-lee/react-ion-slider/compare/1.0.12...1.0.13)

> 22 February 2021

- sync [`#6`](https://github.com/gipyeong-lee/react-ion-slider/pull/6)
- modified release-it config [`#5`](https://github.com/gipyeong-lee/react-ion-slider/pull/5)
- implement release-it [`#4`](https://github.com/gipyeong-lee/react-ion-slider/pull/4)
- fixed prop-types unmet [`a0586ef`](https://github.com/gipyeong-lee/react-ion-slider/commit/a0586ef13a323048b44e061d8ba72c5238035935)
- Release react-ion-slider 1.0.13 [`c3494bb`](https://github.com/gipyeong-lee/react-ion-slider/commit/c3494bb82244fb43e8f385df001852e9b2c3dd7d)
- test#4 [`4742ecb`](https://github.com/gipyeong-lee/react-ion-slider/commit/4742ecb3696a138953bf4a967331ba92c45bd98f)

#### [1.0.12](https://github.com/gipyeong-lee/react-ion-slider/compare/1.0.11...1.0.12)

> 22 February 2021

- implement changeLogs [`b492acb`](https://github.com/gipyeong-lee/react-ion-slider/commit/b492acb71cb4602e50513b6152f45d9d3f423bd7)
- Release 1.0.12 [`1ea1b8c`](https://github.com/gipyeong-lee/react-ion-slider/commit/1ea1b8cc9f883fd9cdf2a618d72542b14ddf5e47)

#### 1.0.11

> 22 February 2021

- Dependabot/npm and yarn/elliptic 6.5.3 [`#1`](https://github.com/gipyeong-lee/react-ion-slider/pull/1)
- Bump mixin-deep from 1.3.1 to 1.3.2 in /example [`#7`](https://github.com/gipyeong-lee/react-ion-slider/pull/7)
- Bump js-yaml from 3.7.0 to 3.13.1 [`#8`](https://github.com/gipyeong-lee/react-ion-slider/pull/8)
- Bump mixin-deep from 1.3.1 to 1.3.2 [`#9`](https://github.com/gipyeong-lee/react-ion-slider/pull/9)
- Bump eslint-utils from 1.3.1 to 1.4.3 [`#10`](https://github.com/gipyeong-lee/react-ion-slider/pull/10)
- Bump lodash.template from 4.4.0 to 4.5.0 in /example [`#15`](https://github.com/gipyeong-lee/react-ion-slider/pull/15)
- Bump acorn from 5.7.3 to 5.7.4 [`#17`](https://github.com/gipyeong-lee/react-ion-slider/pull/17)
- Bump websocket-extensions from 0.1.3 to 0.1.4 in /example [`#18`](https://github.com/gipyeong-lee/react-ion-slider/pull/18)
- Bump websocket-extensions from 0.1.3 to 0.1.4 [`#19`](https://github.com/gipyeong-lee/react-ion-slider/pull/19)
- Bump handlebars from 4.1.1 to 4.5.3 [`#12`](https://github.com/gipyeong-lee/react-ion-slider/pull/12)
- Bump handlebars from 4.1.1 to 4.5.3 in /example [`#13`](https://github.com/gipyeong-lee/react-ion-slider/pull/13)
- update function support [`#6`](https://github.com/gipyeong-lee/react-ion-slider/pull/6)
- implement document [`f87f9ab`](https://github.com/gipyeong-lee/react-ion-slider/commit/f87f9ab855d5d6510a73637ac9fc7c48f051d5cf)
- init create-react-library@2.6.7 [`40fa9d7`](https://github.com/gipyeong-lee/react-ion-slider/commit/40fa9d71538ec1a65c1de0d98cea990f7eee2b88)
- implement release-it [`dae05c0`](https://github.com/gipyeong-lee/react-ion-slider/commit/dae05c0dbf2bedd2f781510fa431e6acdf6190a5)


example/src/App.js
import React, {Component, Fragment} from 'react'

import IonRangeSlider from 'react-ion-slider'

export default class App extends Component {
  state = {
    skin: 'flat',
    min: 10,
    max: 100
  }
  onClickRerender = () => {
    console.log('onClick!!')
    this.setState(Object.assign({}, this.state, {skin: 'modern', min: 50, max: 200}), () => {
      this.ionSlider &&
      this.ionSlider.update({skin: 'flat', min: 100, max: 500, from: 120, to: 240, type: 'double'})
    })
  }

  render() {
    return (
      <Fragment>
        <div onClick={this.onClickRerender}>Re Render</div>
        <div style={{paddingLeft: '20px', width: '100px'}}>
          {/* <IonRangeSlider type={'double'} skin={'flat'} style={{ color: 'white' }} className={'helo'} min={10} max={1000}
            onFinish={(data)=>{console.log(data)}} /> */}
          <IonRangeSlider ref={r => this.ionSlider = r} skin={this.state.skin} min={this.state.min}
                          max={this.state.max} />
        </div>
      </Fragment>
    )
  }
}
